### PR TITLE
Fix ReferenceError when requestAnimationFrame isn't defined

### PR DESCRIPTION
**Pull request by gaearon**
Fixes https://github.com/facebook/react/pull/12980#issuecomment-402600801.

We can't access a global if it's not defined so we should always guard it. I consolidated these reads in one place. I also changed the test to reproduce the problem on master.

### CODE REVIEW

File: packages/react-scheduler/src/ReactScheduler.js [hunk 1]
```diff
@@ -64,12 +52,24 @@ const localDate = Date;
...
+// We don't expect either of these to necessarily be defined,
+// but we will error later if they are missing on the client.
+const localRequestAnimationFrame =
+  typeof requestAnimationFrame === 'function'
+    ? requestAnimationFrame
+    : (undefined: any);
+const localCancelAnimationFrame =
+  typeof cancelAnimationFrame === 'function'
+    ? cancelAnimationFrame
+    : (undefined: any);
```
Comment (acdlite) [inline packages/react-scheduler/src/ReactScheduler.js]: This module throws if this is undefined, right? Maybe we should use an invariant?

**Comment by react-linter** [comment]
ReactDOM: size: :small_red_triangle:+0.1%, gzip: 0.0%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 88d7ed8bfbccd860c3e309da39d356d0a3127aa7...9064cf5d753c8168d806a658c4609083a13413c8</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +0.1% | 0.0% | 638.02 KB | 638.46 KB | 149.34 KB | 149.41 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.1%** | **0.0%** | **95.59 KB** | **95.69 KB** | **30.9 KB** | **30.91 KB** | **UMD_PROD**
react-dom.development.js | +0.1% | 0.0% | 634.15 KB | 634.58 KB | 148.17 KB | 148.23 KB | NODE_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.1%** | **0.0%** | **95.57 KB** | **95.67 KB** | **30.41 KB** | **30.42 KB** | **NODE_PROD**
react-dom.profiling.min.js | +0.1% | 0.0% | 96.54 KB | 96.64 KB | 30.75 KB | 30.76 KB | NODE_PROFILING

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +0.1% | +0.1% | 422.62 KB | 423.06 KB | 95.02 KB | 95.1 KB | UMD_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.1%** | **-0.0%** | **82.71 KB** | **82.81 KB** | **25.47 KB** | **25.47 KB** | **UMD_PROD**
react-art.development.js | +0.1% | +0.1% | 354.6 KB | 355.04 KB | 77.89 KB | 77.96 KB | NODE_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.2%** | **:small_red_triangle:+0.1%** | **47.69 KB** | **47.79 KB** | **14.85 KB** | **14.86 KB** | **NODE_PROD**

## react-scheduler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-scheduler.development.js | +2.7% | +1.5% | 16.3 KB | 16.74 KB | 5.07 KB | 5.15 KB | UMD_DEV
**react-scheduler.production.min.js** | **:small_red_triangle:+3.6%** | **:small_red_triangle:+0.9%** | **2.62 KB** | **2.72 KB** | **1.24 KB** | **1.25 KB** | **UMD_PROD**
react-scheduler.development.js | +2.7% | +1.6% | 16.11 KB | 16.55 KB | 5.03 KB | 5.11 KB | NODE_DEV
**react-scheduler.production.min.js** | **:small_red_triangle:+3.7%** | **:small_red_triangle:+0.7%** | **2.54 KB** | **2.64 KB** | **1.19 KB** | **1.2 KB** | **NODE_PROD**
ReactScheduler-dev.js | +3.3% | +1.6% | 14.87 KB | 15.36 KB | 4.55 KB | 4.63 KB | FB_WWW_DEV
**ReactScheduler-prod.js** | **:small_red_triangle:+1.6%** | **:small_red_triangle:+0.3%** | **7.47 KB** | **7.59 KB** | **1.9 KB** | **1.91 KB** | **FB_WWW_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: packages/react-scheduler/src/ReactScheduler.js (hunk 1 above)
Comment (gaearon) [reply packages/react-scheduler/src/ReactScheduler.js]: Didn't want to spend the bytes.

I think it actually used an invariant in the past, and Sebastian wanted to relax that to a warning.

Comment (acdlite) [reply packages/react-scheduler/src/ReactScheduler.js]: I guess in sync mode this module is never used so an invariant would be harsh

**Comment by bvaughn** [comment]
Nice! thanks for the fix :)

**Pull request merged**